pkg/bcdb/tx_context.go

```
@@ -115,6 +115,15 @@ func (t *commonTxContext) commit(tx txContext, postEndpoint string, sync bool) (
 
 	t.txSpent = true
 	tx.cleanCtx()
+
+	if sync {
+		receipt := txResponseEnvelope.GetResponse().GetReceipt()
+		validationInfo := receipt.GetHeader().GetValidationInfo()
+		if validationInfo != nil && validationInfo[receipt.TxIndex].GetFlag() != types.Flag_VALID {
+			return txID, receipt, errors.Errorf("transaction txID = %s is not valid, reason: %s", txID, validationInfo[receipt.TxIndex].GetFlag().String())
```

If `validationInfo == nil`, it is a server programming error, it should never happen. Should return a different error.

Instead of using `errors.Errorf`, when the tx is invalid, please define a new type that implement `Error() string`, e.g. `ErrorTxValidation`. As an example for that see `&ServerTimeout{TxID: txID}` above. This will enable the user to detect this type of errors using type switch and handle them accordingly (re-executing the tx, for example).